Allow one way partitionings

andersmusikkahs opened the pull request:
This is an attempt to solve #167 .

mcches commented on src/sim.rs:
```diff
@@ -660,6 +683,52 @@ mod test {
+            Ok(())
+        });
+
+        sim.partition_oneway("receiver", "sender");
```
Could you also add the following tests please:

- verify repair works after partition
- check with two different a and b pairings to ensure sort ordering works as intended

andersmusikkahs replied:
I added tests for repair.

I made the test so that it doesn't always create client "a" first, to try to establish that the sort-orderings don't depend on which client is created first, but I'm not sure if this is what you're asking for!

mcches commented on src/sim.rs:
```diff
@@ -660,6 +683,52 @@ mod test {
+        });
+
+        sim.client("sender", async move {
+            // Peers are partitioned. TCP setup should fail.
```
TCP?

andersmusikkahs replied:
Ooops😬! A bit of copy-paste may have happened here.

andersmusikkahs replied:
Reference to TCP removed!

mcches commented on src/top.rs:
```diff
@@ -440,7 +482,55 @@ impl Link {
+        self.state = State::ExplicitPartitionBothWays;
+    }
+
+    fn explicit_partition_a_to_b(&mut self) {
```
We'll also need test coverage over these state transitions.

andersmusikkahs replied:
I added an exhaustive test that all 6 operations that affect partitioning do work as expected in all 6! (=720) combinations.

This gives some peace of mind, but it takes a while to run in debug-mode (about 8 seconds) . One possibility would be to just disable this test for debug-builds. Another possibility would be to not run all 720 combinations.

mcches replied:
I think we could limit the number of cases down to still be pretty fast, but also give us confidence in the implementation. You could also use a lower level test on this module that doesn't need all the packets to move around or all the host scheduling.

andersmusikkahs replied:
Done!

mcches commented on src/top.rs:
```diff
@@ -332,14 +368,20 @@ impl Link {
                 let delay = self.delay(global_config.latency(), rand);
                 DeliveryStatus::DeliverAfter(self.now + delay)
             }
+            State::ExplicitPartitionAToB if src.ip() > dst.ip() => {
```
This is a bit tricky to reason about as written. I assume this delivers in the non-partitioned link direction. Consider adding a comment on these branches.

andersmusikkahs replied:
I agree. I'll see if I can make it clearer.

andersmusikkahs replied:
I've tried to make it clearer.

mcches reviewed (changes requested):
Thanks for picking up this issue!

mcches commented on src/top.rs:
```diff
@@ -155,7 +155,15 @@ enum State {
     Healthy,
 
     /// The link was explicitly partitioned.
-    ExplicitPartition,
```
I haven't seen a need for this yet, but one idea to achieve this is to model a link with two direction states and allow setting either one. This also might make the A,B -> src,dst stuff cleaner. wdyt?

andersmusikkahs replied:
I think it would be a lot cleaner if the 'links'-map in the 'Topology'-struct didn't deduplicate links. I.e, if the network topology between any two nodes 'a' and 'b' was described by two links, 'a'->'b' and 'b'->'a', instead of having one multidirectional link.

However, that feels like a rather big change, and it might also have other downsides.

We could just split 'state' in struct 'Link' into 'state_a_to_b' and 'state_b_to_a'. I worry that the tricky A,B->src,dst stuff would basically remain the same though?

mcches replied:
I agree we should hold off on a larger refactor. We've been discussing a larger restructuring of the network to support different ip versions and a more realistic lo for some time now. We can tackle that then. 

This idea doesn't avoid complexity around the directions, but it does make it really clear what is happening in each direction. I think its worth doing that as part of this change.

```rust
struct Link {
    a_b: State,
    b_a: State,
}
```

andersmusikkahs replied:
Right! I made an attempt, and I have to say it was a big improvement. Thanks for the idea!

mcches reviewed (commented):
This is coming together nicely. I left a few comments, please let me know what you think.

mcches commented on src/sim.rs:
```diff
@@ -661,6 +686,207 @@ mod test {
+        };
+
+        let construct_a_first = sim.world.borrow_mut().rng.gen_bool(0.5);
+        if construct_a_first {
```
What is the significance of the construction order?

mcches commented on src/sim.rs:
```diff
@@ -661,6 +686,207 @@ mod test {
+    }
+
+    #[test]
+    fn partition_peers_oneway_many_cases() -> Result {
```
How's the duration on this now?

andersmusikkahs replied:
It's quite fast.It's about 1/100 of the number of cases compared to before, which really improved performance.

mcches reviewed (commented):
Mainly just one minor comment about arg names. Looks like you also have a couple clippy violations: https://github.com/tokio-rs/turmoil/actions/runs/11164962511/job/31053944943?pr=187

andersmusikkahs commented on src/top.rs:
```diff
@@ -251,10 +259,38 @@ impl Topology {
         self.links[&Pair::new(a, b)].explicit_partition();
     }
 
+    pub(crate) fn partition_oneway(&mut self, a: IpAddr, b: IpAddr) {
```
Yeah, that's obviously much better!

andersmusikkahs replied:
I have implemented your suggestion (I hope :-) ).

mcches commented:
Looks like this needs a rebase and was hoping we could get `from` and `to` on the pub facing Apis for this.

andersmusikkahs commented:
> Looks like this needs a rebase and was hoping we could get `from` and `to` on the pub facing Apis for this.

I rebased, squashed into a single commiit, and renamed the parameter 'a' and 'b' to 'from' and 'to'.

mcches commented on src/sim.rs:
```diff
@@ -216,6 +225,19 @@ impl<'a> Sim<'a> {
+    /// from 'from' to 'to', while not affecting the ability for them to be delivered
+    /// in the other direction.
+    ///
+    /// Partitioning first a->b, then b->a, will stop all messages, similarly to
```
```suggestion
    /// Partitioning first from->to, then to->from, will stop all messages, similarly to
```

mcches reviewed (approved):
I pulled your changes and everything looks good. Thanks for the contribution!
